components/raftstore-v2/src/operation/bucket.rs

```diff
@@ -0,0 +1,200 @@
+            assert_eq!(buckets.len(), 1);
+            let bucket_keys = buckets.pop().unwrap().keys;
+            let bucket_count = bucket_keys.len() + 1;
+
+            let mut meta = BucketMeta {
+                region_id: self.region_id(),
+                region_epoch,
+                version: gen_bucket_version(self.term(), current_version),
+                keys: bucket_keys,
+                sizes: vec![store_ctx.coprocessor_host.cfg.region_bucket_size.0; bucket_count],
+            };
+            // padding the boundary keys and initialize the flow.
+            meta.keys.insert(0, region.get_start_key().to_vec());
+            meta.keys.push(region.get_end_key().to_vec());
+            let stats = new_bucket_stats(&meta);
+            region_buckets = BucketStat::new(Arc::new(meta), stats);
+        }
+
+        let buckets_count = region_buckets.meta.keys.len() - 1;
+        store_ctx.coprocessor_host.on_region_changed(
+            region,
+            RegionChangeEvent::UpdateBuckets(buckets_count),
+            self.state_role(),
+        );
+        let meta = region_buckets.meta.clone();
+        self.set_region_buckets(Some(region_buckets));
+        let mut store_meta = store_ctx.store_meta.lock().unwrap();
+        if let Some(reader) = store_meta.readers.get_mut(&self.region_id()) {
+            reader.0.update(ReadProgress::region_buckets(meta));
+        }
+    }
+
+    #[inline]
+    pub fn report_region_buckets_pd<T>(&mut self, ctx: &StoreContext<EK, ER, T>) {
+        let region_buckets = self.region_buckets().as_ref().unwrap();
+        let task = pd::Task::ReportBuckets(region_buckets.clone());
+        if let Err(e) = ctx.schedulers.pd.schedule(task) {
+            error!(
+                self.logger,
+                "failed to report buckets to pd";
+                "err" => ?e,
+            );
+            return;
+        }
+        self.clear_region_bucket_stats();
+    }
+
+    pub fn maybe_gen_approximate_buckets<T>(&self, ctx: &StoreContext<EK, ER, T>) {
+        if ctx.coprocessor_host.cfg.enable_region_bucket && !self.region().get_peers().is_empty() {
+            if let Err(e) = ctx
+                .schedulers
+                .split_check
+                .schedule(SplitCheckTask::ApproximateBuckets(self.region().clone()))
+            {
+                error!(
+                    self.logger,
+                    "failed to schedule check approximate buckets";
+                    "err" => %e,
+                );
+            }
+        }
+    }
+}
+
+impl<'a, EK, ER, T: Transport> PeerFsmDelegate<'a, EK, ER, T>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    #[inline]
+    pub fn on_report_region_buckets_tick(&mut self) {
+        if self.fsm.peer().is_leader() && self.fsm.peer().region_buckets().is_some() {
+            self.fsm.peer_mut().report_region_buckets_pd(self.store_ctx);
+        }
+        self.schedule_tick(PeerTick::ReportBuckets);
```

Comment:
Why schedule the tick when it's follower?

Reply:
yes, I will remove it and add tick if the peer become leader or bucker refreshed.